### ISSUE TITLE
Use `http_archive` with private Gitlab repo

### ISSUE THREAD

**Issue by kilian-funk** [bazelbuild/bazel#16659]
### Description of the feature request:

I would like to use this in my WORKSPACE file:

```
http_archive(
    name = "my_repo",
    url = "https://gitlab.com/kilian-funk/test_project/-/archive/main/test_project-main.tar.gz",
    auth_patterns = {"gitlab.com": "PRIVATE-TOKEN: <password>"},
}
```
where `test_project` is a private repository.

### What underlying problem are you trying to solve with this feature?

The tarball containing the repo can be downloaded on the command line with:
```
curl --header "PRIVATE-TOKEN: <token-goes-here>" <url>
```
Unfortunately the `http_archive` rule results in something equivalent to:
```
curl --header "Authentication: PRIVATE-TOKEN: <token-goes-here>" <url>
```
Therefore, `http_archive` is not compatible with gitlab.com for private repos, since it does not let the user customize the authentication header in the required way. `http_archive` is making too many assumptions on how an authentication header should look like.

### Which operating system are you running Bazel on?

Linux, Ubuntu

### What is the output of `bazel info release`?

release 5.3.2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

---

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
---
```

### Have you found anything relevant by searching the web?

no

### Any other information, logs, or outputs that you want to share?

-

**Comment by Wyverald** [comment]
Sending a custom HTTP header is a highly uncommon use case. For GitLab specifically, you can use the standard Authentication header instead: https://docs.gitlab.com/ee/api/#personalprojectgroup-access-tokens

**Comment by Wyverald** [comment]
any links for context?

**Comment by thesayyn** [comment]
https://github.com/opencontainers/distribution-spec/blob/main/spec.md#workflow-categories

Part that starts with `The client SHOULD include an Accept header indicating which manifest content types it supports.`

Which is currently can't be set; bazel automatically sets it to `text/html, image/gif, image/jpeg, */*`

**Comment by Wyverald** [comment]
@thesayyn I'm much more sympathetic towards adding an "Accept" header than a completely random one called "PRIVATE-TOKEN". Could you file a separate issue about this? Thanks.

**Comment by thesayyn** [comment]
> @thesayyn I'm much more sympathetic towards adding an "Accept" header than a completely random one called "PRIVATE-TOKEN". Could you file a separate issue about this? Thanks.

filed https://github.com/bazelbuild/bazel/issues/17829. I'll cc you

**Comment by tjgq** [comment]
FYI - you should be able to do this using a [credential helper](https://github.com/bazelbuild/proposals/blob/main/designs/2022-06-07-bazel-credential-helpers.md) in Bazel 6.3.0 (soon to be released) or later.

**Comment by thesayyn** [comment]
> FYI - you should be able to do this using a [credential helper](https://github.com/bazelbuild/proposals/blob/main/designs/2022-06-07-bazel-credential-helpers.md) in Bazel 6.3.0 (soon to be released) or later.

The problem with credential_helper is that I have to tell my ruleset users to either put a bash script into their workspace or download a binary from the internet for every CI/Dev machine just to add a header. it's not bazel-like

**Comment by CauhxMilloy** [comment]
I recently ran into this issue, but found out a solution. I'm responding here just to go ahead and document it (hopefully also helping anyone else who happens to stumble upon this in the future).

It is true that gitlab has historically used `"PRIVATE-TOKEN: <access token>"` as their header for their REST APIs / HTTP requests. However, as of 2019, the APIs now also accept the more common Oauth-style header, `Authorization: Bearer <access token>`. This can be seen [here](https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/api/rest/index.md#personalprojectgroup-access-tokens) in their documentation, and it was added in [this commit](https://gitlab.com/gitlab-org/gitlab/-/commit/aba93fe2d5661cf3c086f65838db2965c746fdbf).

This can be shown working using `curl -v --header "Authorization: Bearer $MY_ACCESS_TOKEN" "$MY_HTTP_REQUEST_URL"`.

This can be used with Bazel with:
```
http_archive(
    name = "my_external_repo",
    auth_patterns = {
        "gitlab.com": "Bearer <password>"
    },
    netrc = "/path/to/my/.netrc/if/not/wanting/to/use/the/one/in/my/home/dir/.netrc",
    sha256 = MY_EXT_REPO_SHA,
    url = MY_EXT_REPO_URL,
)
```
(one could also use the `NETRC` environment variable in lieu of using the `netrc` attribute.)

Note: the netrc file must be given in an absolute path. Relative paths are not supported; they resolve to the `external/` directory in the cache. This is probably not where your `.netrc` file is -- or anything else that you'd really want to reference (maybe should be addressed) 😇.

Hope this helps!
